### PR TITLE
Add Dump command to VS

### PR DESCRIPTION
**Pull request by AviAvni**
image
Talking with @dsyme I implemented Dump command to vs to help anlyze memory issues of F# language service in VS
The dump taken after all operations in Reactor finished
And I used ClrMD https://github.com/Microsoft/clrmd/pull/91 to take the dump so it not waste disk or ram
The dump written to the output window
@cartermp what do you think?
I'm also want to implement this on tests so we can compare memory after every build

One thing missing is thaat even that I added the Microsoft.Diagnostics.Runtime package to FSharp.Editor
and VisualFSharpFull projects the dll is not included in the vsix how I can fix this?

### CODE REVIEW

**Comment by majocha** [comment]
@AviAvni adding to VisualFSharpFull.csproj 
```
  <ItemGroup>
      <VSIXSourceItem Include="$(OutputPath)Microsoft.Diagnostics.Runtime.dll" />
  </ItemGroup>
```


should probably work, but I have no idea if this is the recommended way to do it.

File: vsintegration/src/FSharp.ProjectSystem.FSharp/MenusAndCommands.vsct [hunk 1]
```diff
@@ -133,6 +133,15 @@
           <ButtonText>F# Interactive</ButtonText>
         </Strings>
       </Button>
+      
+      <Button guid="FSharpProjectCmdSet" id="cmdidFsDump" priority="0x8000" type="Button">
```
Comment (brettfo) [inline vsintegration/src/FSharp.ProjectSystem.FSharp/MenusAndCommands.vsct]: This addition should have (but obviously didn't) update all of the `MenusAndCommands.vsct.*.xlf` files.  Can you try running this to force regeneration?

``` bash
msbuild vsintegration\src\FSharp.ProjectSystem.FSharp\ProjectSystem.fsproj /t:UpxateXlf
```

File: vsintegration/Vsix/VisualFSharpFull/VisualFSharpFull.csproj [hunk 2]
```diff
@@ -189,6 +189,7 @@
 
   <ItemGroup>
     <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonPackageVersion)" />
+    <PackageReference Include="Microsoft.Diagnostics.Runtime" Version="0.9.180305.1" />
```
Comment (brettfo) [inline vsintegration/Vsix/VisualFSharpFull/VisualFSharpFull.csproj]: This version number should be added to `build\targets\PackageVersions.props` under the `other packages` heading and placed in alphabetical order:

``` xml
<MicrosoftDiagnosticsRuntimePackageVersion>0.9.180305.1</MicrosoftDiagnosticsRuntimePackageVersion>
```

And replaced in this file with `$(MicrosoftDiagnosticsRuntimePackageVersion)`.

File: vsintegration/src/FSharp.Editor/Commands/DumpCommandService.fs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+[<TextViewRole(PredefinedTextViewRoles.PrimaryDocument)>]
+type internal DumpCommandFilterProvider 
+    [<ImportingConstructor>] 
+    (checkerProvider: FSharpCheckerProvider,
+     [<Import(typeof<SVsServiceProvider>)>] serviceProvider: System.IServiceProvider) =
+
+    let projectSystemPackage =
+      lazy(
+        let shell = serviceProvider.GetService(typeof<SVsShell>) :?> IVsShell
+        let packageToBeLoadedGuid = ref (Guid "{91a04a73-4f2c-4e7c-ad38-c1a68e7da05c}")  // FSharp ProjectSystem guid
```
Comment (brettfo) [inline vsintegration/src/FSharp.Editor/Commands/DumpCommandService.fs]: Use the constant defined `ProjectPrelude.fs` as `GuidList.guidFSharpProjectPkgString`.

File: vsintegration/src/FSharp.Editor/Commands/DumpCommandService.fs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        match shell.LoadPackage packageToBeLoadedGuid with
+        | VSConstants.S_OK, pkg ->
+            pkg :?> Package
+        | _ -> null)
+
+    let FSharpDump (_this:Package) (_sender:obj) (_e:EventArgs) =
+        checkerProvider.Checker.ClearLanguageServiceRootCachesAndCollectAndFinalizeAllTransients()
+        use target = DataTarget.CreateSnapshotAndAttach(Process.GetCurrentProcess().Id)
+        let runtime = target.ClrVersions.[0].CreateRuntime();
+        let outputPane = projectSystemPackage.Value.GetOutputPane(VSConstants.OutputWindowPaneGuid.BuildOutputPane_guid, "FSharp Dump")
```
Comment (brettfo) [inline vsintegration/src/FSharp.Editor/Commands/DumpCommandService.fs]: I'm not familiar with this API.  Does "FSharp Dump" need to be localized/translated?  If so it'll have to be added to a `.resx` file and consumed that way.

File: vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj [hunk 5]
```diff
@@ -139,6 +141,7 @@
     <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonPackageVersion)" />
     <PackageReference Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutablePackageVersion)" />
     <PackageReference Include="VSSDK.VSLangProj" Version="$(VSSDKVSLangProjPackageVersion)" />
+    <PackageReference Include="Microsoft.Diagnostics.Runtime" Version="0.9.180305.1" />
```
Comment (brettfo) [inline vsintegration/src/FSharp.Editor/FSharp.Editor.fsproj]: Same `build\targets\PackageVersions.props` commend as above.

**Comment by Pilchie** [comment]
Sorry for the late reply - was on vacation.  I would tend to prefer that this be added as an extension, perhaps into the "Project System Tools" extension?  Tagging @panopticoncentral

**Comment by AviAvni** [comment]
@brettfo and @Pilchie  thanks for feedback
This is in the "Project System Tools" because @dsyme wanted it in a specific state of the language service when all the work on the queue is done
In any case I'll be happy to continue work on this

**Comment by AviAvni** [comment]
@panopticoncentral ok so how this is going to work?
can I use classes that declared in this repo there via nuget so I can take the dump when the lang service finished all operations
or I'll just do the dump there and it's need to be taken manually on time
or I'll expose it as a service and use that service here
or you have another idea?

**Comment by panopticoncentral** [comment]
Is the API to know when the lang service is done available publicly? If so, I'd just call it from the extension and then do the dump.

**Comment by KevinRansom** [comment]
@AviAvni 

Hello my friend, is this still an active work item for you?

Thanks
Kevin

**Comment by AviAvni** [comment]
I can work on it if you want but as said in comments it belong to other project that I don't know how to implement it there

**Comment by KevinRansom** [comment]
Thanks for this PR, however it looks to be quite out of date, please reopen, if you want to continue with it.

Kevin

**Pull request closed without merging**